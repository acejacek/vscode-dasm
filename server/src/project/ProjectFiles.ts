import {
	Position,
	Range,
	TextDocument,
} from "vscode-languageserver";

import * as fs from "fs";
import * as path from "path";
import SimpleSignal from "simplesignal";

import PathUtils from "../utils/PathUtils";
import StringUtils from "../utils/StringUtils";

export interface IProjectFileDependency {
	parentRelativeUri: string;
	range: Range;
	file?: IProjectFile;
}

export interface IProjectFile {
	uri: string;
	document?: TextDocument;
	contents?: string;
	contentsLines?: string[];
	dependencies: IProjectFileDependency[];
	isDirty: boolean;
	version: number;
}

interface IFileUpdateEvent {
	uri: string;
	process: () => void;
}

export default class ProjectFiles {

	private static readonly FILE_UPDATE_DEBOUNCE_MS = 250; // Time to wait until a file change is propagated

	private _files:{[key:string]:IProjectFile};
	private _entryFile?:IProjectFile;
	private _onChanged:SimpleSignal<(files:ProjectFiles) => void>;
	private _onAdded:SimpleSignal<(files:ProjectFiles) => void>;

	private _queuedFileUpdate?:IFileUpdateEvent;
	private _queuedUpdateProcessId?:NodeJS.Timer;

	constructor() {
		this._files = {};
		this._entryFile = undefined;
		this._onChanged = new SimpleSignal<(files:ProjectFiles) => void>();
		this._onAdded = new SimpleSignal<(files:ProjectFiles) => void>();
	}

	public addFromDocument(document:TextDocument) {
		this.addByUri(document.uri);
		this.updateFromDocument(document);
	}

	public addByUri(uri:string) {
		// Create new document if one doesn't exist
		const cleanUri = this.getCleanUri(uri);
		if (!this.has(uri)) {
			const newFile:IProjectFile = {
				uri: cleanUri,
				document: undefined,
				contents: undefined,
				contentsLines: undefined,
				dependencies: [],
				isDirty: false,
				version: -1,
			};

			if (!this._entryFile) {
				this._entryFile = newFile;
			}

			this._files[cleanUri] = newFile;
			this._onAdded.dispatch(this);
		}
	}

	public debug_logProject() {
		const keys = Object.keys(this._files);
		console.log("  " + keys.length + " files: ");
		keys.forEach((key, index) => {
			const file = this._files[key];
			console.log(`    ${index}: ${file.uri}`);
			console.log(`      has document = ${Boolean(file.document)}`);
			console.log(`      has content = ${Boolean(file.contents)}, length = ${file.contents ? file.contents.length : -1}`);
			console.log(`      is dirty = ${Boolean(file.isDirty)}`);
			console.log(`      version = ${file.version}`);
			console.log(`      includes = ${file.dependencies.length}`);
		});
	}

	public updateFromDocument(document:TextDocument) {
		const fileInfo = this.get(document.uri);
		if (fileInfo && fileInfo.version !== document.version) {
			this.queueFileUpdate(document.uri, () => {
				// Needs update
				fileInfo.document = document;
				fileInfo.contents = document.getText();
				fileInfo.contentsLines = StringUtils.splitIntoLines(fileInfo.contents);
				fileInfo.version = document.version;
				fileInfo.isDirty = true;

				this.updateDependencies(fileInfo);
				this._onChanged.dispatch(this);
			});
		}
	}

	public updateFromFileSystem(uri:string) {
		console.log("[files] Adding file as uri", uri);
		const fileInfo = this.get(uri);
		if (fileInfo) {
			this.queueFileUpdate(uri, () => {
				// Needs update
				fileInfo.document = undefined;
				fileInfo.contents = fs.readFileSync(PathUtils.uriToPlatformPath(uri), {encoding: "utf8"});
				fileInfo.contentsLines = StringUtils.splitIntoLines(fileInfo.contents);
				fileInfo.version = -1;
				fileInfo.isDirty = false;

				this.updateDependencies(fileInfo);
				this._onChanged.dispatch(this);
			});
		}
	}

	public markSaved(document:TextDocument) {
		const fileInfo = this.get(document.uri);
		if (fileInfo) fileInfo.isDirty = false;
	}

	public markClosed(document:TextDocument) {
		const fileInfo = this.get(document.uri);
		if (fileInfo) fileInfo.document = undefined;
	}

	public has(uri:string) {
		return this._files.hasOwnProperty(this.getCleanUri(uri));
	}

	public get(uri:string):IProjectFile|undefined {
		const cleanUri = this.getCleanUri(uri);
		return this.has(cleanUri) ? this._files[cleanUri] : undefined;
	}

	public getEntry():IProjectFile|undefined {
		return this._entryFile;
	}

	public all():IProjectFile[] {
		const files:IProjectFile[] = [];
		const keys = Object.keys(this._files);
		keys.forEach((key) => {
			if (this.has(key)) files.push(this.get(key) as IProjectFile);
		});
		return files;
	}

	public getByDependencyUri(parentUri:string|undefined, parentRelativeUri:string):IProjectFile|undefined {
		let possibleParents:IProjectFile[] = [];
		if (parentUri) {
			// Parent-specific dependency
			const parent = this.get(parentUri);
			if (parent) {
				possibleParents = [parent];
			}
		} else {
			// Any parent
			possibleParents = this.all();
		}

		for (const parent of possibleParents) {
			const dependency = parent.dependencies.find((dependencyInfo) => dependencyInfo.parentRelativeUri === parentRelativeUri);
			if (dependency) return dependency.file;
		}

		return undefined;
	}

	public get onAdded() {
		return this._onAdded;
	}

	public get onChanged() {
		return this._onChanged;
	}

	/**
	 * Creates a list of all includes, with the entry-relative uri as the key and the contents as the value
	 */
	public getIncludes():{[key:string]:string} {
		return this._entryFile ? this.buildIncludes(this._entryFile) : {};
	}

	public getSource():string|undefined {
		return this._entryFile ? this._entryFile.contents : undefined;
	}

	public dispose() {
		this.clearQueuedFileUpdate();
		this._onAdded.removeAll();
		this._onChanged.removeAll();
	}

	private getCleanUri(uri:string):string {
		return decodeURIComponent(uri);
	}

	/**
	 * Trigger an event to update all dependencies,
	 * with proper debouncing so it doesn't do that too often
	 */
	private queueFileUpdate(uri:string, process:() => void) {
		// Cancel existing update event if existing
		this.clearQueuedFileUpdate();

		// If an event exists and it's a different file, process it immediately
		if (this._queuedFileUpdate && this._queuedFileUpdate.uri !== uri) {
			this._queuedFileUpdate.process();
		}

		// Finally, add the new one as the next one
		this._queuedFileUpdate = {
			uri,
			process,
		};

		// And queue it to execute next
		this._queuedUpdateProcessId = setTimeout(() => {
			process();
			this._queuedUpdateProcessId = undefined;
		}, ProjectFiles.FILE_UPDATE_DEBOUNCE_MS);
	}

	private clearQueuedFileUpdate() {
		if (this._queuedUpdateProcessId) {
			clearTimeout(this._queuedUpdateProcessId);
			this._queuedUpdateProcessId = undefined;
		}
	}

	/**
	 * Return a list of all files included within a source, with their include filename and range
	 */
	private getIncludedFileLinks(lines:string[]) {
		const includeFind = /^[^;\n]+include\s*([^ ;\n]*)/gmi;
		const files:Array<{fileName:string, range:Range}> = [];
		lines.forEach((line, lineIndex) => {
			let result = includeFind.exec(line);
			while (result) {
				const fileName:string = StringUtils.removeWrappingQuotes(result[1]);
				if (fileName) files.push({
					fileName,
					range: Range.create(
						Position.create(lineIndex, result.index + result[0].indexOf(result[1])),
						Position.create(lineIndex, result.index + result[0].length),
					),
				});
				result = includeFind.exec(line);
			}
		});
		return files;
	}

	/**
	 * Given a file info, update its existing dependencies, removing
	 * the ones that are not there and adding the new ones
	 */
	private updateDependencies(file:IProjectFile) {
		if (file.contentsLines) {
			const allDependencyLinks = this.getIncludedFileLinks(file.contentsLines);

			// Remove dependencies that are not featured anymore
			file.dependencies = file.dependencies.filter(
				(dependencyInfo) => allDependencyLinks.some(
					(dependencyLink) => dependencyLink.fileName === dependencyInfo.parentRelativeUri,
				),
			);

			// Filter down to new dependencies only
			const newDependencyLinks = allDependencyLinks.filter((dependencyLink) => {
				return !file.dependencies.some((dependencyInfo) => dependencyInfo.parentRelativeUri === dependencyLink.fileName);
			});
			const newDependencies:IProjectFileDependency[] = newDependencyLinks.map((dependencyLink) => {
				return {
					parentRelativeUri: dependencyLink.fileName,
					range: dependencyLink.range,
				};
			});
			file.dependencies = file.dependencies.concat(newDependencies);

			// TODO: this is a bit messy, to get an uri like a real file uri. need to check on osx, or make it more platform agnostic

			// Add dependency files to the main file list where needed
			const baseFolder = PathUtils.uriToPlatformPath(path.dirname(file.uri));
			file.dependencies.forEach((dependencyInfo) => {
				const uri = PathUtils.platformPathToUri(path.join(baseFolder, dependencyInfo.parentRelativeUri));
				if (!this.has(uri)) {
					// Doesn't exist: a new file, add to the project
					this.addByUri(uri);
					this.updateFromFileSystem(uri);
				}

				// Update its contents
				dependencyInfo.file = this.get(uri);
			});
		}
	}

	private buildIncludes(file:IProjectFile, previousRelativeUri?:string) {
		const includes:{[key:string]:string} = {};
		file.dependencies.forEach((dependency) => {
			// Includes the file itself
			const includeUri = previousRelativeUri ? path.join(previousRelativeUri, dependency.parentRelativeUri) : dependency.parentRelativeUri;
			if (dependency.file && dependency.file.contents) {
				includes[includeUri] = dependency.file.contents;
				// Includes its own dependencies
				Object.assign(includes, this.buildIncludes(dependency.file, includeUri));
			} else {
				console.error(`Error! Trying to include "${previousRelativeUri}" as a file dependency without contents!`);
			}
		});
		return includes;
	}
}

