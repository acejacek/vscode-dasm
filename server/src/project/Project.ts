import {
	TextDocument,
} from "vscode-languageserver";

import { Assembler, IAssemblerResult } from "./../providers/Assembler";
import ProjectFiles from "./ProfileFiles";

export default class Project {

	private _files:ProjectFiles;
	private _results?:IAssemblerResult;
	private _assembler:Assembler;

	constructor() {
		this._assembler = new Assembler();

		this._files = new ProjectFiles();
		this._files.onChanged.add(this.assemble.bind(this));

		this._results = undefined;
	}

	public addFile(document:TextDocument) {
		this._files.addFromDocument(document);
	}

	public debug_logProject() {
		this._files.debug_logProject();
	}

	public updateFile(document:TextDocument) {
		this._files.updateFromDocument(document);
	}

	public markFileSaved(document:TextDocument) {
		this._files.markSaved(document);
	}

	public markFileClosed(document:TextDocument) {
		this._files.markClosed(document);
	}

	public hasFile(uri:string) {
		return this._files.has(uri);
	}

	public getResults():IAssemblerResult|undefined {
		return this._results;
	}

	public getFileInfo(uri:string) {
		return this._files.get(uri);
	}

	public getFileInfoLocalUri(parentRelativeUri:string) {
		return this._files.getByDependencyUri(undefined, parentRelativeUri);
	}

	private assemble() {
		// Gather all files and compile
		const source = this._files.getSource();
		const includes = this._files.getIncludes();
		if (source && includes) {
			console.time("[project] Assembling");
			this._results = this._assembler.assemble(source, includes);
			console.timeEnd("[project] Assembling");
		} else {
			console.warn("[project] No entry file to compile");
		}
	}
}

