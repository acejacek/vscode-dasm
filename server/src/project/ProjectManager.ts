import {
	IConnection,
	InitializeResult,
	TextDocument,
	TextDocuments,
} from "vscode-languageserver";

import { IAssemblerResult } from "../providers/Assembler";
import CompletionProvider from "../providers/CompletionProvider";
import DefinitionProvider from "../providers/DefinitionProvider";
import DiagnosticsProvider from "../providers/DiagnosticsProvider";
import DocumentLinkProvider from "../providers/DocumentLinkProvider";
import DocumentSymbolProvider from "../providers/DocumentSymbolProvider";
import HoverProvider from "../providers/HoverProvider";
import SettingsProvider from "../providers/SettingsProvider";
import { ISettings } from "../providers/SettingsProvider";
import Project from "./Project";
import { IProjectFile } from './ProjectFiles';

export default class ProjectManager {

	private _connection:IConnection;
	private _projects:Project[];
	private _currentProject?:Project;
	private _currentDocumentUri?:string;
	private _workspaceRoot:string;
	private _documents:TextDocuments;

	private _diagnosticsProvider:DiagnosticsProvider;
	private _hoverProvider:HoverProvider;
	private _definitionProvider:DefinitionProvider;
	private _settingsProvider:SettingsProvider;
	private _completionProvider:CompletionProvider;
	private _documentLinkProvider:DocumentLinkProvider;
	private _documentSymbolProvider:DocumentSymbolProvider;

	constructor(connection:IConnection) {
		this._connection = connection;
		this._projects = [];
		this._currentProject = undefined;
		this._currentDocumentUri = undefined;

		// Initialize hooks; full document sync (open, change and close document events)
		this._documents = new TextDocuments();
		this._documents.listen(this._connection);

		// Initialization
		this._connection.onInitialize((params):InitializeResult => {
			this._workspaceRoot = params.rootUri || "";

			return {
				// Tells the client about the server's capabilities
				capabilities: {
					// Working in FULL text document sync mode
					textDocumentSync: this._documents.syncKind,

					// Hover on symbols/etc
					hoverProvider: true,

					// Code complete
					completionProvider: {
						resolveProvider: true,
						triggerCharacters: [ "." ],
					},

					// Go to definition
					definitionProvider: true,

					// Links (dependencies)
					documentLinkProvider: {
						resolveProvider: true,
					},

					// Symbols per document
					documentSymbolProvider: true,
				},
			};
		});

		// When a text document is opened or when its content has changed
		this._documents.onDidOpen((change) => {
			this.debug_logProjects();
			this.onDocumentOpened(change.document);
		});

		this._documents.onDidChangeContent((change) => {
			this.debug_logProjects();
			this.onDocumentChanged(change.document);
		});

		this._documents.onDidSave((change) => {
			this.debug_logProjects();
			this.onDocumentSaved(change.document);
		});

		this._documents.onDidClose((change) => {
			this.debug_logProjects();
			this.onDocumentClosed(change.document);
		});

		// Create providers
		const projectInfoProvider = {
			getEntryFiles: this.getEntryFiles.bind(this),
			getFile: this.getFile.bind(this),
			getResults: this.getCurrentResults.bind(this),
			getUriForProjectFile: this.getUriForProjectFile.bind(this),
			getSettings: this.getSettings.bind(this),
		};

		this._diagnosticsProvider = new DiagnosticsProvider(this._connection, projectInfoProvider);
		this._hoverProvider = new HoverProvider(this._connection, projectInfoProvider);
		this._definitionProvider = new DefinitionProvider(this._connection, projectInfoProvider);
		this._settingsProvider = new SettingsProvider(this._connection, projectInfoProvider);
		this._completionProvider = new CompletionProvider(this._connection, projectInfoProvider);
		this._documentLinkProvider = new DocumentLinkProvider(this._connection, projectInfoProvider);
		this._documentSymbolProvider = new DocumentSymbolProvider(this._connection, projectInfoProvider);
	}

	public start() {
		// Listen on the connection
		this._connection.listen();
	}

	private debug_logProjects() {
		console.log("---");
		console.log(" ALL tabs: ", this._documents.keys());
		console.log(" Projects: ", this._projects.length);
		if (this._currentProject) {
			console.log(" CURRENT project:");
			this._currentProject.debug_logProject();
		}
		// this._projects.forEach((project, index) => {
		// 	console.log(index + " ----");
		// 	project.debug_logProject();
		// });
		console.log("---");
	}

	/**
	 * A documented was opened; fired before a change event when switched to a new tab
	 */
	private onDocumentOpened(document:TextDocument) {
		this.setCurrentDocument(document);
	}

	/**
	 * A document's contents was updated
	 */
	private onDocumentChanged(document:TextDocument) {
		this.setCurrentDocument(document);

		// Re-assemble the current project
		if (this._currentProject) {
			this._currentProject.updateFile(document);
			this.updatePostProviders();
		}
	}

	/**
	 * Set a document as the current active document tab (and current "project")
	 */
	private setCurrentDocument(document:TextDocument) {
		if (document.uri !== this._currentDocumentUri) {
			// Tab changed: change context, changing (and creating) a new project if needed
			console.log("[pm] NEW CURRENT URI is ", document.uri);
			const project = this.getProjectForFile(document);

			if (project && project !== this._currentProject) {
				// A new project altogether, set it as default
				console.log("[pm] ...ALSO SETTING new project");
				this._currentProject = project;
			}

			this._currentDocumentUri = document.uri;
		}
	}

	/**
	 * After closing a document, don't know the current one
	 */
	private unsetCurrentDocument() {
		this._currentDocumentUri = undefined;
		this._currentProject = undefined;
	}

	/**
	 * A documented was saved
	 */
	private onDocumentSaved(document:TextDocument) {
		this.setCurrentDocument(document);

		if (this._currentProject) {
			this._currentProject.markFileSaved(document);
			this.updatePostProviders();
		}
	}

	/**
	 * A documented was closed
	 */
	private onDocumentClosed(document:TextDocument) {
		if (this._currentProject) {
			this._currentProject.updateFile(document);
			this._currentProject.markFileClosed(document);
		}
		// TODO: close project if all files are closed

		this.unsetCurrentDocument();
	}

	/**
	 * Updates existing providers with the latest info from the current tab
	 */
	private updatePostProviders() {
		const fileUri = this._currentDocumentUri;
		if (fileUri) {
			// Diagnostics
			this._diagnosticsProvider.process(fileUri);
		}
	}

	/**
	 * Get a list of what is considered the "entry" file for all currently known projects
	 */
	private getEntryFiles():IProjectFile[] {
		const allEntryFiles = this._projects.map((projectInfo) => projectInfo.getEntryFileInfo());
		return allEntryFiles.filter((file) => Boolean(file)) as IProjectFile[];
	}

	/**
	 * Based on an URI, find a file in any of the current projects
	 */
	private getFile(uri:string):IProjectFile|undefined {
		const project = this._projects.find((projectInfo) => projectInfo.hasFile(uri));
		if (project) return project.getFileInfo(uri);
	}

	private getCurrentResults():IAssemblerResult|undefined {
		if (this._currentProject) {
			return this._currentProject.getResults();
		}
	}

	private getUriForProjectFile(parentRelativeUri:string) {
		if (this._currentProject) {
			const file = this._currentProject.getFileInfoLocalUri(parentRelativeUri);
			return file ? file.uri : undefined;
		}
	}

	/**
	 * Returns the project a file belongs to
	 * TODO: Some files (e.g. include files) can belong to more than one project!
	 */
	private getProjectForFile(document:TextDocument, avoidCreating:boolean = false):Project|undefined {
		let newProject = this._projects.find((project) => project.hasFile(document.uri));
		if (!newProject && !avoidCreating) {
			// A new file, create a project for it
			newProject = new Project();
			newProject.addFile(document);
			this._projects.push(newProject);
		}
		return newProject;
	}

	private getSettings():ISettings {
		return this._settingsProvider.getCurrent();
	}
}
