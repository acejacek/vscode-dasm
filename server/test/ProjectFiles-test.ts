import * as path from "path";
import * as url from "url";

import ProjectFiles from "../src/project/ProjectFiles";
import { IProjectFile } from "../src/project/ProjectFiles";

function createProjectFiles() {
	const pf = new ProjectFiles();
	pf.debounceTime = 0;
	return pf;
}

function clearFileUri(fileUri: string) {
	const filePath = url.parse(fileUri).pathname;
	const rootPath = path.posix.sep + __dirname.split(path.sep).splice(1).join(path.posix.sep);
	if (filePath) {
		return "<test-path>/" + path.posix.relative(rootPath, filePath);
	} else {
		return "?";
	}
}

function clearFileContent(contents?: string) {
	return `Redacted (was ${contents ? `${contents.length} chars in ${contents.split("\n").length} lines` : "undefined"})`;
}

function clearFile(file: IProjectFile): IProjectFile {
	const newContent = clearFileContent(file.contents);
	return {
		...file,
		uri: clearFileUri(file.uri),
		dependencies: file.dependencies.map((dependency) => {
			return {
				...dependency,
				uri: clearFileUri(dependency.uri),
				file: dependency.file ? clearFile(dependency.file) : undefined,
			};
		}),
		contents: newContent,
		contentsLines: [ newContent ],
	};
}

function getIncludeList(includes?: {[key:string]:string}) {
	if (includes) {
		const keys = Object.keys(includes);
		const newIncludes:{[key:string]:string} = {};
		for (const key of keys) {
			newIncludes[key] = clearFileContent(includes[key]);
		}
		return newIncludes;
	} else {
		return undefined;
	}
}

describe("test adding with includes", () => {
	test("is empty", () => {
		const pf = createProjectFiles();
		expect(pf.has("a")).toBe(false);
		expect(pf.all().length).toEqual(0);
	});

	test("no-include project", () => {
		const pf = createProjectFiles();

		const filePath = path.posix.join(__dirname, "sources/_test_single.asm");
		pf.addFromDiskPath(filePath);

		expect(pf.all().length).toEqual(1);
		expect(pf.all().map(clearFile)).toMatchSnapshot();

		const includes = getIncludeList(pf.getIncludes());
		expect(includes).toMatchSnapshot();
		expect(Object.keys(includes).length).toBe(0);
	});

	test("single-include project", () => {
		const pf = createProjectFiles();

		const filePath = path.posix.join(__dirname, "sources/_test_include.asm");
		pf.addFromDiskPath(filePath);

		expect(pf.all().length).toEqual(2);
		expect(pf.all().map(clearFile)).toMatchSnapshot();

		const includes = getIncludeList(pf.getIncludes());
		expect(includes).toMatchSnapshot();
		expect(Object.keys(includes).length).toBe(1);
	});

	test("single-include project with incdir", () => {
		const pf = createProjectFiles();

		const filePath = path.posix.join(__dirname, "sources/_test_include_incdir.asm");
		pf.addFromDiskPath(filePath);

		expect(pf.all().length).toEqual(2);
		expect(pf.all().map(clearFile)).toMatchSnapshot();

		const includes = getIncludeList(pf.getIncludes());
		expect(includes).toMatchSnapshot();
		expect(Object.keys(includes).length).toBe(1);
	});

	test("single-include project with missing include", () => {
		const pf = createProjectFiles();

		const filePath = path.posix.join(__dirname, "sources/_test_broken.asm");
		pf.addFromDiskPath(filePath);

		expect(pf.all().length).toEqual(2);
		expect(pf.all().map(clearFile)).toMatchSnapshot();

		const includes = getIncludeList(pf.getIncludes());
		expect(includes).toMatchSnapshot();
		expect(Object.keys(includes).length).toBe(0);
	});
});
