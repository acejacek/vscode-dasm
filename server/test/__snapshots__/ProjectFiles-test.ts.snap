// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test adding with includes no-include project 1`] = `
Array [
  Object {
    "contents": "Redacted (was 20856 chars in 499 lines)",
    "contentsLines": Array [
      "Redacted (was 20856 chars in 499 lines)",
    ],
    "dependencies": Array [],
    "document": undefined,
    "isDirty": false,
    "uri": "<test-path>/sources/_test_single.asm",
    "version": -1,
  },
]
`;

exports[`test adding with includes no-include project 2`] = `Object {}`;

exports[`test adding with includes single-include project 1`] = `
Array [
  Object {
    "contents": "Redacted (was 11683 chars in 300 lines)",
    "contentsLines": Array [
      "Redacted (was 11683 chars in 300 lines)",
    ],
    "dependencies": Array [
      Object {
        "file": Object {
          "contents": "Redacted (was 9189 chars in 201 lines)",
          "contentsLines": Array [
            "Redacted (was 9189 chars in 201 lines)",
          ],
          "dependencies": Array [],
          "document": undefined,
          "isDirty": false,
          "uri": "<test-path>/sources/vcs.h",
          "version": -1,
        },
        "parentRelativeUri": "vcs.h",
        "range": Object {
          "end": Object {
            "character": 15,
            "line": 18,
          },
          "start": Object {
            "character": 10,
            "line": 18,
          },
        },
        "uri": "<test-path>/sources/vcs.h",
      },
    ],
    "document": undefined,
    "isDirty": false,
    "uri": "<test-path>/sources/_test_include.asm",
    "version": -1,
  },
  Object {
    "contents": "Redacted (was 9189 chars in 201 lines)",
    "contentsLines": Array [
      "Redacted (was 9189 chars in 201 lines)",
    ],
    "dependencies": Array [],
    "document": undefined,
    "isDirty": false,
    "uri": "<test-path>/sources/vcs.h",
    "version": -1,
  },
]
`;

exports[`test adding with includes single-include project 2`] = `
Object {
  "vcs.h": "Redacted (was 9189 chars in 201 lines)",
}
`;

exports[`test adding with includes single-include project with incdir 1`] = `
Array [
  Object {
    "contents": "Redacted (was 11759 chars in 303 lines)",
    "contentsLines": Array [
      "Redacted (was 11759 chars in 303 lines)",
    ],
    "dependencies": Array [
      Object {
        "file": Object {
          "contents": "Redacted (was 9189 chars in 201 lines)",
          "contentsLines": Array [
            "Redacted (was 9189 chars in 201 lines)",
          ],
          "dependencies": Array [],
          "document": undefined,
          "isDirty": false,
          "uri": "<test-path>/sources/includes/includes_vcs.h",
          "version": -1,
        },
        "parentRelativeUri": "includes/includes_vcs.h",
        "range": Object {
          "end": Object {
            "character": 24,
            "line": 21,
          },
          "start": Object {
            "character": 10,
            "line": 21,
          },
        },
        "uri": "<test-path>/sources/includes/includes_vcs.h",
      },
    ],
    "document": undefined,
    "isDirty": false,
    "uri": "<test-path>/sources/_test_include_incdir.asm",
    "version": -1,
  },
  Object {
    "contents": "Redacted (was 9189 chars in 201 lines)",
    "contentsLines": Array [
      "Redacted (was 9189 chars in 201 lines)",
    ],
    "dependencies": Array [],
    "document": undefined,
    "isDirty": false,
    "uri": "<test-path>/sources/includes/includes_vcs.h",
    "version": -1,
  },
]
`;

exports[`test adding with includes single-include project with incdir 2`] = `
Object {
  "includes/includes_vcs.h": "Redacted (was 9189 chars in 201 lines)",
}
`;

exports[`test adding with includes single-include project with missing include 1`] = `
Array [
  Object {
    "contents": "Redacted (was 3670 chars in 188 lines)",
    "contentsLines": Array [
      "Redacted (was 3670 chars in 188 lines)",
    ],
    "dependencies": Array [
      Object {
        "file": Object {
          "contents": "Redacted (was undefined)",
          "contentsLines": Array [
            "Redacted (was undefined)",
          ],
          "dependencies": Array [],
          "document": undefined,
          "isDirty": true,
          "uri": "<test-path>/sources/a/vcs_wrong.h",
          "version": -1,
        },
        "parentRelativeUri": "a/vcs_wrong.h",
        "range": Object {
          "end": Object {
            "character": 23,
            "line": 28,
          },
          "start": Object {
            "character": 10,
            "line": 28,
          },
        },
        "uri": "<test-path>/sources/a/vcs_wrong.h",
      },
    ],
    "document": undefined,
    "isDirty": false,
    "uri": "<test-path>/sources/_test_broken.asm",
    "version": -1,
  },
  Object {
    "contents": "Redacted (was undefined)",
    "contentsLines": Array [
      "Redacted (was undefined)",
    ],
    "dependencies": Array [],
    "document": undefined,
    "isDirty": true,
    "uri": "<test-path>/sources/a/vcs_wrong.h",
    "version": -1,
  },
]
`;

exports[`test adding with includes single-include project with missing include 2`] = `Object {}`;
